### PR TITLE
EnC: Check member modifier updates in semantic analysis

### PR DESCRIPTION
**Pull request by tmat**
Analyzing modifier changes in semantic analysis has multiple advantages: 
1) We can detect the actual impact of change on the generated code. In cases when changing a modifier does not affect the generated code we don't need to issue a rude edit.
2) When analyzing syntax of a type or its member without semantic information we do not know whether to report a rude edit or not - the type may be reloadable, in which case most rude edits are not applicable.

The PR also fixes a few issues uncovered by the changes in the analyzer.

### CODE REVIEW

**Comment by tmat** [comment]
@davidwengier

**Comment by tmat** [comment]
@davidwengier Seems like we are missing tests that update attributes on non-type members (specifically events).

**Comment by davidwengier** [comment]
> Seems like we are missing tests that update attributes on non-type members (specifically events).

Thanks. Logged https://github.com/dotnet/roslyn/issues/54430

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 1]
```diff
@@ -2888,18 +2884,19 @@ public ConstructorEdit(INamedTypeSymbol oldType)
...
                                         if (isConstructorWithMemberInitializers)
                                         {
                                             processedSymbols.Remove(newSymbol);
                                         }
 
-                                        // Need to check for attribute rude edits for fields and properties
-                                        AnalyzeCustomAttributes(oldSymbol, newSymbol, capabilities, diagnostics, semanticEdits, syntaxMap, cancellationToken);
+                                        if (isDeclarationWithInitializer)
+                                        {
+                                            AnalyzeSymbolUpdate(oldSymbol, newSymbol, newDeclaration, capabilities, diagnostics, semanticEdits, syntaxMap, cancellationToken);
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:2898]: I'm sure this is covered by some other case (and hopefully tests), but this looks like an issue. Whether there is an initializer, we still need to report rude edits for attributes etc., right?

**Review by davidwengier — APPROVED** [review]
The actual modifier update bits looks good. The analyzer changes are difficult to reason about without debugging through, but I trust the tests are doing the right thing, so I'm sure it's fine. Plus you're _fairly_ trustworthy too 😛

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 1 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:2898]: We call DeferConstructor edit for consturctors that emit member initializers and declarations with initializers. This takes care of handling symbol changes (attributes, modifiers) for the later. The former is handled in AddConstructorEdits where we actually issue edits for the constructors.

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs [hunk 2]
```diff
@@ -1867,7 +1867,7 @@ class C
             var active = GetActiveStatements(src1, src2);
 
             edits.VerifyRudeDiagnostics(active,
-                Diagnostic(RudeEditKind.ModifiersUpdate, "const int a = 1", FeaturesResources.const_field));
+                Diagnostic(RudeEditKind.ModifiersUpdate, "a = 1", FeaturesResources.const_field));
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs:1870]: ❓ why does this code change from `const int a = 1` to  `a = 1`? is the `const int` no longer important for the test?

File: src/EditorFeatures/CSharpTest/EditAndContinue/Helpers/CSharpEditAndContinueTestHelpers.cs [hunk 3]
```diff
@@ -11,6 +11,7 @@
 using Microsoft.CodeAnalysis.EditAndContinue;
 using Microsoft.CodeAnalysis.EditAndContinue.UnitTests;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest/EditAndContinue/Helpers/CSharpEditAndContinueTestHelpers.cs:14]: Is this used?

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs (hunk 2 above)
Comment (tmat) [reply src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs:1870]: It's the span of the squiggle. The syntax reference associated with the symbol is different from the syntax used by the previous analysis.

**Pull request merged**